### PR TITLE
Source Facebook Marketing: Fix fail when async job takes too long

### LINKED ISSUE #7723: Source Facebook Marketing fails when async job takes too long
## Current Behavior
Confirmed on the latest version of facebook marketing

Source FB marketing source fails if the async job takes too long to run with error `} did not finish after 2049 seconds. This is an intermittent error which may be fixed by retrying the job. Aborting.` (full logs are below) 

Notice that this job **was not failing** but rather running, it just seems to be taking a long time. 

## Expected Behavior
As long as the sync job is making progress (which this one seems to be doing) the connector shouldn't fail. Either that or it should 

Some ideas for a solution: 
1. increase wait time (not very reliable since a job can take long)
2. decrease job size so long running jobs are not very likely
3. use a different heuristic to identify if a job is "making progress" -- maybe saying something like if the percentage completion hasn't moved in 20 minutes then it's stuck, and retry the job instead

## Logs
<details>
<summary>LOG</summary> 

```
[34msource[0m - 2021-11-05 14:52:28 WARN DefaultAirbyteStreamFactory(internalLog):96 - '0' is not of type 'null', 'number'
[34msource[0m - 2021-11-05 15:26:39 ERROR DefaultAirbyteStreamFactory(internalLog):95 - Encountered an exception while reading stream SourceFacebookMarketing
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/sources/abstract_source.py", line 104, in read
    internal_config=internal_config,
  File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/sources/abstract_source.py", line 134, in _read_stream
    for record in record_iterator:
  File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/sources/abstract_source.py", line 179, in _read_incremental
    for record_counter, record_data in enumerate(records, start=1):
  File "/airbyte/integration_code/source_facebook_marketing/streams.py", line 330, in read_records
    result = self.wait_for_job(stream_slice["job"])
  File "/usr/local/lib/python3.7/site-packages/backoff/_sync.py", line 94, in retry
    ret = target(*args, **kwargs)
  File "/airbyte/integration_code/source_facebook_marketing/streams.py", line 382, in wait_for_job
    f"AdReportRun {job} did not finish after {runtime.in_seconds()} seconds."
source_facebook_marketing.common.JobTimeoutException: AdReportRun <AdReportRun> {
    "account_id": "106960666117991",
    "async_percent_completion": 90,
    "async_status": "Job Running",
    "date_start": "2019-08-26",
    "date_stop": "2019-08-26",
    "id": "4515697388466598",
    "is_running": true,
    "report_run_id": "4515697388466598",
    "time_ref": 1636123511
} did not finish after 2049 seconds. This is an intermittent error which may be fixed by retrying the job. Aborting.
[34msource[0m - 2021-11-05 15:26:39 ERROR LineGobbler(voidCall):82 - Traceback (most recent call last):
[34msource[0m - 2021-11-05 15:26:39 ERROR LineGobbler(voidCall):82 -   File "/airbyte/integration_code/main.py", line 13, in <module>
[34msource[0m - 2021-11-05 15:26:39 ERROR LineGobbler(voidCall):82 -     launch(source, sys.argv[1:])
[34msource[0m - 2021-11-05 15:26:39 ERROR LineGobbler(voidCall):82 -   File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/entrypoint.py", line 108, in launch
[34msource[0m - 2021-11-05 15:26:39 ERROR LineGobbler(voidCall):82 -     for message in source_entrypoint.run(parsed_args):
[34msource[0m - 2021-11-05 15:26:39 ERROR LineGobbler(voidCall):82 -   File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/entrypoint.py", line 99, in run
[34msource[0m - 2021-11-05 15:26:39 ERROR LineGobbler(voidCall):82 -     for message in generator:
[34msource[0m - 2021-11-05 15:26:39 ERROR LineGobbler(voidCall):82 -   File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/sources/abstract_source.py", line 108, in read
[34msource[0m - 2021-11-05 15:26:39 ERROR LineGobbler(voidCall):82 -     raise e
[34msource[0m - 2021-11-05 15:26:39 ERROR LineGobbler(voidCall):82 -   File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/sources/abstract_source.py", line 104, in read
[34msource[0m - 2021-11-05 15:26:39 ERROR LineGobbler(voidCall):82 -     internal_config=internal_config,
[34msource[0m - 2021-11-05 15:26:39 ERROR LineGobbler(voidCall):82 -   File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/sources/abstract_source.py", line 134, in _read_stream
[34msource[0m - 2021-11-05 15:26:39 ERROR LineGobbler(voidCall):82 -     for record in record_iterator:
[34msource[0m - 2021-11-05 15:26:39 ERROR LineGobbler(voidCall):82 -   File "/usr/local/lib/python3.7/site-packages/airbyte_cdk/sources/abstract_source.py", line 179, in _read_incremental
[34msource[0m - 2021-11-05 15:26:39 ERROR LineGobbler(voidCall):82 -     for record_counter, record_data in enumerate(records, start=1):
[34msource[0m - 2021-11-05 15:26:39 ERROR LineGobbler(voidCall):82 -   File "/airbyte/integration_code/source_facebook_marketing/streams.py", line 330, in read_records
[34msource[0m - 2021-11-05 15:26:39 ERROR LineGobbler(voidCall):82 -     result = self.wait_for_job(stream_slice["job"])
[34msource[0m - 2021-11-05 15:26:39 ERROR LineGobbler(voidCall):82 -   File "/usr/local/lib/python3.7/site-packages/backoff/_sync.py", line 94, in retry
[34msource[0m - 2021-11-05 15:26:39 ERROR LineGobbler(voidCall):82 -     ret = target(*args, **kwargs)
[34msource[0m - 2021-11-05 15:26:39 ERROR LineGobbler(voidCall):82 -   File "/airbyte/integration_code/source_facebook_marketing/streams.py", line 382, in wait_for_job
[34msource[0m - 2021-11-05 15:26:39 ERROR LineGobbler(voidCall):82 -     f"AdReportRun {job} did not finish after {runtime.in_seconds()} seconds."
[34msource[0m - 2021-11-05 15:26:39 ERROR LineGobbler(voidCall):82 - source_facebook_marketing.common.JobTimeoutException: AdReportRun <AdReportRun> {
[34msource[0m - 2021-11-05 15:26:39 ERROR LineGobbler(voidCall):82 -     "account_id": "106960666117991",
[34msource[0m - 2021-11-05 15:26:39 ERROR LineGobbler(voidCall):82 -     "async_percent_completion": 90,
[34msource[0m - 2021-11-05 15:26:39 ERROR LineGobbler(voidCall):82 -     "async_status": "Job Running",
[34msource[0m - 2021-11-05 15:26:39 ERROR LineGobbler(voidCall):82 -     "date_start": "2019-08-26",
[34msource[0m - 2021-11-05 15:26:39 ERROR LineGobbler(voidCall):82 -     "date_stop": "2019-08-26",
[34msource[0m - 2021-11-05 15:26:39 ERROR LineGobbler(voidCall):82 -     "id": "4515697388466598",
[34msource[0m - 2021-11-05 15:26:39 ERROR LineGobbler(voidCall):82 -     "is_running": true,
[34msource[0m - 2021-11-05 15:26:39 ERROR LineGobbler(voidCall):82 -     "report_run_id": "4515697388466598",
[34msource[0m - 2021-11-05 15:26:39 ERROR LineGobbler(voidCall):82 -     "time_ref": 1636123511
[34msource[0m - 2021-11-05 15:26:39 ERROR LineGobbler(voidCall):82 - } did not finish after 2049 seconds. This is an intermittent error which may be fixed by retrying the job. Aborting.

```

</details>

### PR DESCRIPTION
**Pull request by htrueman**
## What
Closes #7723

## How
Expand job retry logics to retry on any job fail.

## Recommended reading order
1. `streams.py`

## Pre-merge Checklist
Expand the relevant checklist and delete the others. 

<details><summary> <strong> New Connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/SUMMARY.md`
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
</p>
</details>

<details><summary> <strong> Updating a connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the new connector version is published, connector version bumped in the seed directory as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</p>
</details>

<details><summary> <strong> Connector Generator </strong> </summary>
<p>
   
- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed.
</p>
</details>

### CODE REVIEW

File: airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/streams.py [hunk 1]
```diff
@@ -369,31 +369,33 @@ def wait_for_job(self, job, stream_state: Mapping[str, Any] = None) -> AdReportR
...
-                    f" This is an intermittent error which may be fixed by retrying the job. Aborting."
-                )
-            elif runtime > self.MAX_WAIT_TO_FINISH:
-                raise JobTimeoutException(
-                    f"AdReportRun {job} did not finish after {runtime.in_seconds()} seconds."
-                    f" This is an intermittent error which may be fixed by retrying the job. Aborting."
-                )
+                else:
+                    if job["async_status"] == "Job Failed":
+                        raise JobTimeoutException(f"AdReportRun {job} failed after {runtime.in_seconds()} seconds.")
```
Comment (avida) [inline airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/streams.py:385]: Please do not use JobTimeoutException here, this error is not related to timeout and very confusing.

File: airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/streams.py [hunk 2]
```diff
@@ -369,31 +369,33 @@ def wait_for_job(self, job, stream_state: Mapping[str, Any] = None) -> AdReportR
...
+                    elif job["async_status"] == "Job Skipped":
+                        raise JobTimeoutException(f"AdReportRun {job} skipped after {runtime.in_seconds()} seconds.")
```
Comment (avida) [inline airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/streams.py:387]: Same here, its not a timeout error

File: airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/streams.py [hunk 3]
```diff
@@ -369,31 +369,33 @@ def wait_for_job(self, job, stream_state: Mapping[str, Any] = None) -> AdReportR
 
             if job["async_status"] == "Job Completed":
                 return job
-            elif job["async_status"] in ["Job Failed", "Job Skipped"]:
+            else:
                 time_range = (job["date_start"], job["date_stop"])
-                if self.times_job_restarted.get(time_range, 0) < 6:
+                self.times_job_restarted[time_range] = self.times_job_restarted.get(time_range, 0)
+                if self.times_job_restarted[time_range] < 6:
```
Comment (avida) [inline airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/streams.py:375]: please move 6 to const variable

File: airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/streams.py [hunk 4]
```diff
@@ -369,31 +369,33 @@ def wait_for_job(self, job, stream_state: Mapping[str, Any] = None) -> AdReportR
 
             if job["async_status"] == "Job Completed":
                 return job
-            elif job["async_status"] in ["Job Failed", "Job Skipped"]:
+            else:
                 time_range = (job["date_start"], job["date_stop"])
-                if self.times_job_restarted.get(time_range, 0) < 6:
+                self.times_job_restarted[time_range] = self.times_job_restarted.get(time_range, 0)
```
Comment (avida) [inline airbyte-integrations/connectors/source-facebook-marketing/source_facebook_marketing/streams.py:374]: you can use defaultdict(int)

**Review by avida — CHANGES_REQUESTED** [review]

**Review by Zirochkaa — APPROVED** [review]
Agree with all @avida comments.

**Review by avida — APPROVED** [review]

**Review by Zirochkaa — APPROVED** [review]

**Review by sherifnada — APPROVED** [review]

**Pull request merged**